SCAutolib/models/card.py

```diff
@@ -0,0 +1,229 @@
+"""
+This module implements classes for communication with different types of cards
+that we are using in the library. Those types are: virtual smart card, real
+(physical) smart card in standard reader, cards in the removinator.
+"""
+import re
+import time
+from pathlib import Path
+from traceback import format_exc
+
+from SCAutolib import run, logger, TEMPLATES_DIR
+from SCAutolib.models.file import SoftHSM2Conf
+from SCAutolib.models.user import User
+
+
+class Card:
+    """
+    Interface for child classes. All child classes will rewrite common methods
+    based on the type of the card.
+    """
+    uri: str = None
+    user: User = None
+    _pattern: str = None
+
+    def _set_uri(self):
+        """
+        Sets URI for given smart card. Uri is matched from ``p11tool`` command
+        with regular expression. If URI is not matched, assertion exception
+        would be raised
+
+        :raise: AssertionError
+        """
+        cmd = ["p11tool", "--list-token-urls"]
+        out = run(cmd).stdout
```

Comment:
As suggested previously by Ivan elsewhere the return code should almost always be checked. Instead changing it here, I suggest to change behavior of run function to check return code by default.

Reply:
done